File: resources/[local]/rpc/client/cl_main.lua
local Resource, Promises, CallIdentifier = GetCurrentResourceName(), {}, 0

RPC = {}

function ClearPromise(callID)
    Citizen.SetTimeout(5000, function()
        Promises[callID] = nil
    end)
end

function ParamPacker(...)
    local params, pack = {...} , {}

    for i = 1, 15, 1 do
        pack[i] = {param = params[i]}
    end

    return pack
end

function ParamUnpacker(params, index)
    local idx = index or 1

    if idx <= #params then
        return params[idx]["param"], ParamUnpacker(params, idx + 1)
    end
end

function UnPacker(params, index)
    local idx = index or 1

    if idx <= 15 then
        return params[idx], UnPacker(params, idx + 1)
    end
end

function TriggerNetEvent(pEvent, ...)
    local payload = msgpack.pack({...})

    if payload:len() < 5000 then
        TriggerServerEventInternal(pEvent, payload, payload:len())
    else
        TriggerLatentServerEventInternal(pEvent, payload, payload:len(), 128000)
    end
end

------------------------------------------------------------------
--                  (Trigger Server Calls)
------------------------------------------------------------------

function RPC.execute(name, ...)
    local callID, solved = CallIdentifier, false
    CallIdentifier = CallIdentifier + 1

    Promises[callID] = promise:new()

    TriggerServerEvent("rpc:request", Resource, name, callID, ParamPacker(...))

    Citizen.SetTimeout(20000, function()
        if not solved then
            Promises[callID]:resolve({nil})
            TriggerServerEvent("rpc:server:timeout", name)
        end
    end)

    local response = Citizen.Await(Promises[callID])

    solved = true

    ClearPromise(callID)

    return ParamUnpacker(response)
end

RegisterNetEvent("rpc:response")
AddEventHandler("rpc:response", function(origin, callID, ...)
    if Resource == origin and Promises[callID] then
        Promises[callID]:resolve(...)
    end
end)

File: resources/[local]/rpc/client/cl_main.js
let Promises = {};
let CallIdentifier = 0;
const Resource = GetCurrentResourceName();

let RPC = {};

const ClearPromise = (callID) => {
  setTimeout(() => {
    Promises[callID] = undefined;
  }, 5000);
};

const Packer = (...params) => {
  let pack = {};
  for (let i = 0; i < 15; i++) {
    pack[i] = { param: params[i] };
  }
  return pack;
};

const UnPacker = (params) => {
  let shit = [];
  for (let i = 0; i < params.length; i++) {
    shit.push(params[i].param);
  }
  return shit;
};

RPC.execute = async (name, ...args) => {
  let callID = CallIdentifier;
  let solved = false;
  CallIdentifier = CallIdentifier + 1;

  let resolve, reject;
  let prom = new Promise((pResolve, pReject) => {
    (resolve = pResolve), (reject = pReject);
  });
  Promises[callID] = { prom: prom, resolve: resolve, reject: reject };

  TriggerServerEvent("rpc:request", Resource, name, callID, Packer(args));
  setTimeout(() => {
    if (!solved) {
      Promises[callID].resolve({ nil });
      TriggerServerEvent("rpc:server:timeout", name);
    }
  }, 20000);
  let response = await Promises[callID].prom;
  console.log(`here: ${JSON.stringify(response)}`);
  solved = true;

  ClearPromise(callID);

  console.log(JSON.stringify(UnPacker(response)));
  return UnPacker(response);
};

RegisterNetEvent("rpc:response");
onNet("rpc:response", (origin, callID, ...args) => {
  if (Resource === origin && Promises[callID]) {
    Promises[callID].resolve(...args);
  }
});
